integration/querydsl/blaze-persistence-querydsl-testsuite/src/test/java/com/blazebit/persistence/querydsl/BasicQueryTest.java

```diff
@@ -0,0 +1,744 @@
+            QDocument sub2 = new QDocument("sub2");
+            BlazeJPAQuery<Tuple> query = new BlazeJPAQuery<Document>(entityManager, cbf).from(document)
+                    .select(document.name.as("documentName"), document.name.substring(0, 2))
+                    .where(document.id.in(select(sub.id).from(sub).where(sub.id.in(select(sub2.id).from(sub2).where(sub2.id.eq(sub.id)))).limit(5)));
+
+            List<Tuple> fetch = query.fetch();
+            assertFalse(fetch.isEmpty());
+        });
+    }
+
+    @Test
+    public void testTransformBlazeJPAQuery() {
+        doInJPA(entityManager -> {
+            Map<Long, String> blep = new BlazeJPAQuery<Document>(entityManager, cbf).from(document)
+                    .where(document.name.length().gt(1))
+                    .groupBy(document.id, document.name)
+                    .transform(GroupBy.groupBy(document.id).as(document.name));
+
+            document.getRoot();
+            assertFalse(blep.isEmpty());
+        });
+    }
+
+    @Test
+    public void testAssociationJoin() {
+        doInJPA(entityManager -> {
+            Map<Person, List<Document>> booksByAuthor = new BlazeJPAQuery<Document>(entityManager, cbf)
+                    .from(person)
+                    .innerJoin(person.ownedDocuments, document)
+                    .transform(GroupBy.groupBy(person).as(GroupBy.list(document)));
+
+            assertNotNull(booksByAuthor);
+        });
+    }
+
+    // NOTE: Entity joins are only supported on Hibernate 5.1+
+    @Test
+    @Category({ NoDatanucleus.class, NoEclipselink.class, NoOpenJPA.class, NoHibernate42.class, NoHibernate43.class, NoHibernate50.class })
+    public void testEntityJoin() {
+        Assume.assumeTrue(jpaProvider.supportsEntityJoin());
+
+        doInJPA(entityManager -> {
+            QPerson otherAuthor = new QPerson("otherAuthor");
+            QDocument otherBook = new QDocument("otherBook");
+            Map<Person, List<Document>> booksByAuthor = new BlazeJPAQuery<Document>(entityManager, cbf)
+                    .from(otherAuthor)
+                    .innerJoin(otherBook).on(otherBook.owner.eq(otherAuthor))
+                    .transform(GroupBy.groupBy(otherAuthor).as(GroupBy.list(otherBook)));
+
+            assertNotNull(booksByAuthor);
+        });
+    }
+
+
+    @Test
+    @Category({ NoDatanucleus.class, NoEclipselink.class, NoOpenJPA.class })
+    public void testFromValues() {
+        doInJPA(entityManager -> {
+            Document theBook = new Document();
+            theBook.setId(1337L);
+            theBook.setName("test");
+
+            List<Document> fetch = new BlazeJPAQuery<Document>(entityManager, cbf)
+                    .fromValues(document, Collections.singleton(theBook))
+                    .select(document)
+                    .fetch();
+
+            assertNotNull(fetch);
+        });
+    }
+
+    @Test
+    @Category({ NoDatanucleus.class, NoEclipselink.class, NoOpenJPA.class })
```

Comment:
Add a comment like e.g. `// NOTE: No advanced sql support for datanucleus, eclipselink and openjpa yet`